### DISCUSSION TITLE
Cache returns objects with missing fields when cached type is a combination of an array and an object

### DISCUSSION THREAD

**Discussion by andoco** [TanStack/query#3659] [General] [answered]
I'm fetching an array of objects using an API client generated by the [Typechain](https://github.com/dethcrypto/TypeChain) library which returns objects of this form:

```typescript
export type FooStructOutput = [
  BigNumber,
  string,
] & { id: BigNumber; name: string };
```

The inital call to `useQuery("foos", () => getFoos());` succeeds and returns an array of objects with all the correct fields.

A second call that returns the cached data returns an array with items that do not have the `id` or `name` fields. So it appears that the cache doesn't like the combination of an array and an object in a single type.

I've also tried querying the cache directly using `queryClient.getQueryCache().find("foos")` to verify that the cache is returning data with missing fields.

Is there any config setting or alternative cache that might support this data type?

Thanks.

**Comment by TkDodo** [comment, marked as answer]
I think you'd need to turn off `structuralSharing`, as this requires JSON compatible structures. Additionally, you could provide your own `isDataEqual` method

**Reply by andoco** [reply]
The `structuralSharing` setting worked. Thanks!
